src/RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs

```diff
@@ -0,0 +1,104 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using RoslynAnalyzer;
+using System;
+using System.Collections.Immutable;
+
+namespace ILTrimmingAnalyzer
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class RequiresUnreferencedCodeAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "IL2026";
+
+        private const string s_trimmingEnabledString = "PublishTrimmed";
+
+        private static readonly LocalizableString s_title = new LocalizableResourceString(
+            nameof(RequiresUnreferencedCodeAnalyzer) + "Title",
+            Resources.ResourceManager,
+            typeof(Resources));
+        private static readonly LocalizableString s_messageFormat = new LocalizableResourceString(
+            nameof(RequiresUnreferencedCodeAnalyzer) + "Message",
+            Resources.ResourceManager,
+            typeof(Resources));
+        private const string s_category = "Trimming";
+
+        private static DiagnosticDescriptor s_rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            s_title,
+            s_messageFormat,
+            s_category,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var isTrimmingEnabled = context.Options.GetMSBuildPropertyValue(
+                    s_trimmingEnabledString, compilation, context.CancellationToken);
+
+                if (!string.Equals(isTrimmingEnabled?.Trim(), "true", StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(operationContext =>
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+                    var method = invocation.TargetMethod;
+                    var attributes = method.GetAttributes();
+
+                    foreach (var attr in attributes)
+                    {
+                        if (attr.AttributeClass is { } attrClass &&
+							IsNamedType(attrClass, "System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute") &&
+							attr.ConstructorArguments.Length == 1 &&
+							attr.ConstructorArguments[0] is { Type: { SpecialType: SpecialType.System_String } } ctorArg)
+                        {
+							operationContext.ReportDiagnostic (Diagnostic.Create (
```

Comment:
As the first prototype this is good. Eventually we will need to make sure that if the caller has the attribute, then it doesn't generate warnings in its body, for this attribute.